D/T markers on cache icons

ztNFny opened the pull request:
Adds (optional) D/T markers to cache icons. Controlled by a setting in "Map" preferences, disabled by default, as discussed in yesterdays call.
Also moves the "owned" marker from upper-right to upper-left (fixes #12315)

![image](https://user-images.githubusercontent.com/1258173/145866006-e5bd7f63-f68c-4a82-87c8-a5cf0f2a29fb.png)

fixes #12312

fm-sys commented on main/res/values/strings.xml:
```diff
@@ -826,6 +826,8 @@
     <string name="init_bigSmileysOnMap">Big icons</string>
     <string name="init_summary_bigSmileysOnMap">If enabled, symbols like the \'found smiley\' or the \'corrected coordinates\' marker will be shown enlarged on the map instead of the cache type icon (requires restart).</string>
+    <string name="init_dtMarkerOnCacheIcon">D/T rating on cache icons</string>
+    <string name="init_summary_dtMarkerOnCacheIcon">Show an marker for the Difficulty and Terrain ratings on cache markers in upper-right corner ("stored" marker is moved to middle) (requires restart).</string>
```
Can we add a small sentence to the summary which half of the circle belongs to which of both values?

ztNFny replied:
done

eddiemuc commented on main/src/cgeo/geocaching/utils/MapMarkerUtils.java:
```diff
@@ -697,4 +705,41 @@ public static void resetLists() {
+
+        final String packageName = CgeoApplication.getInstance().getPackageName();
+
+        Map<String, Integer> dt = new HashMap<>();
```
Really necessary to create a map here with two entries? Might it be possible to extract the remainder of the method into a submethod and call it twice instead? This might increase readability

eddiemuc commented on main/src/cgeo/geocaching/utils/MapMarkerUtils.java:
```diff
@@ -697,4 +705,41 @@ public static void resetLists() {
+        dt.put("t", (int) terrain * 10);
+
+        for (Map.Entry<String, Integer> p: dt.entrySet()) {
+            if (p.getValue() >= 10) {
```
I assume that the value is < 10 for caches without d/t rating (e.g. labs) correct? What happens in this case? Is an empty drawable created?

ztNFny replied:
The drawable is created with 2 empty sections as I didn't want to create an empty "0" drawable.

ztNFny replied:
Had to add them now though to be eliminate the map ...

eddiemuc reviewed (approved):
Looks ok codewise. Made some non-critical comments. Thanks for the nice feature!

ztNFny commented:
I replaced the HashMap for drawing the sections with a sub-function, had to add 4 new drawables for that though.
Also added some input sanitation in case some obscure service ever gets the idea of using ratings like 1,25

ztNFny commented:
Empty ratings look like this:
![image](https://user-images.githubusercontent.com/1258173/145870392-14a36c77-5961-4546-931d-1c94bd2b1f4e.png)

fm-sys commented on main/src/cgeo/geocaching/utils/MapMarkerUtils.java:
```diff
@@ -91,6 +94,8 @@ public static CacheMarker getCacheMarker(final Resources res, final Geocache cac
             .append(showFloppyOverlay(cacheListType))
             .append(assignedMarkers)
+            .append(cache.getTerrain())
+            .append(cache.getDifficulty())
```
If we add more and more parameters to the HashMap, does caching makes sense at all? Would it make sense to add the DT value only to the HashMap, if the related setting is enabled?

ztNFny replied:
Caching definitly makes sense. A "standard" situation of viewing a list, then that list as map, then previewing the cache, then opening it would result in 4 times drawing the marker instead of one. So even if the icon is unique it'll still be beneficial.

I'll make those 2 criteria optional though as you suggested.

fm-sys commented:
> Also added some input sanitation in case some obscure service ever gets the idea of using ratings like 1,25

E.g. Extremcaching allows D/T value between 1-8. Can this be handled without a crash (we don't need special drawables for it though)?

Lineflyer commented:
What if the circles would be filled from top to down? This would increase the visibility of lower D/T ratings as the color within the circle would then be more distant and not overlapping the cache type icon. I tried to mock it up on the upper icon in that screenshot:
 
![image](https://user-images.githubusercontent.com/949669/145871454-12203a71-a9e7-4426-b533-cc6389af2843.png)

ztNFny commented:
> E.g. Extremcaching allows D/T value between 1-8. Can this be handled without a crash (we don't need special drawables for it though)?

Yes, "r" will always be an int between 0 and 50 in steps of 5. A rating of 1,2 would result in 1, 8 in 5.

ztNFny commented:
> What if the circles would be filled from top to down? This would increase the visibility of lower D/T ratings as the color within the circle would then be more distant and not overlapping the cache type icon. I tried to mock it up on the upper icon in that screenshot:

I tried that, liked the bottom-filling better. Like a glass of water fills from the bottom, just feels more right to me.

ztNFny commented on main/src/cgeo/geocaching/utils/MapMarkerUtils.java:
```diff
@@ -697,4 +705,39 @@ public static void resetLists() {
+        final int r = Math.max(0, Math.min((int) (Math.round(rating * 2) / 2) * 10, 50));
+        final int tintColor = ResourcesCompat.getColor(res, res.getIdentifier("marker_rating_" + r, "color", packageName), null);
+        final Drawable segment = DrawableCompat.wrap(ResourcesCompat.getDrawable(res, res.getIdentifier("marker_rating_" + ratingLetter + "_" + r, "drawable", packageName), null));
+        DrawableCompat.setTint(segment, tintColor);
```
> Or don't we, because every segment can only be tinted in one specific color?

Yes, we don't :-)

eddiemuc commented:
> Empty ratings look like this:
> ![image](https://user-images.githubusercontent.com/1258173/145870392-14a36c77-5961-4546-931d-1c94bd2b1f4e.png)
> 

What about showing no circle at all on caches without an rating? (Just an idea, and maybe it even makes the code easier... I would also be fine with leaving it as is)

ztNFny commented:
> What if the circles would be filled from top to down?

Btw, my first attempt at this was using 1 icon for d and t each, filling them like a clock:
![image](https://user-images.githubusercontent.com/1258173/145877269-6d5c7739-3825-4c3a-95b3-a7a1b4e875d9.png)
(composed image with all layers visible)

While that works great for 1 rating there just isn't enough room for 2 markers.

ztNFny commented:
I've moved the center for the segments slightly from the middle to the left/right to account for the room lost to the divider bar. This increases the size relative size of 10 and 15 markers slightly.
![image](https://user-images.githubusercontent.com/1258173/146036495-562dbb41-27ff-4580-aa29-8067257b772e.png)
